File: python/infinicore/ops/deepseek_v4_rmsnorm_self.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def deepseek_v4_rmsnorm_self(input: Tensor, epsilon: float, *, out: Tensor | None = None) -> Tensor:
    if out is None:
        return Tensor(_infinicore.deepseek_v4_rmsnorm_self(input._underlying, epsilon))
    _infinicore.deepseek_v4_rmsnorm_self_(out._underlying, input._underlying, epsilon)
    return out


def deepseek_v4_rmsnorm_self_(out: Tensor, input: Tensor, epsilon: float) -> Tensor:
    _infinicore.deepseek_v4_rmsnorm_self_(out._underlying, input._underlying, epsilon)
    return out

File: src/infinicore/deepseek_v4_ops/deepseek_v4_rmsnorm_self/deepseek_v4_rmsnorm_self.cc
#include "infinicore/ops/deepseek_v4_rmsnorm_self.hpp"

namespace infinicore::op {

Tensor deepseek_v4_rmsnorm_self(const Tensor &x, float epsilon) {
    auto out = Tensor::empty(x->shape(), x->dtype(), x->device());
    deepseek_v4_rmsnorm_self_kernel_(out, x, epsilon);
    return out;
}

void deepseek_v4_rmsnorm_self_(Tensor out, const Tensor &x, float epsilon) {
    deepseek_v4_rmsnorm_self_kernel_(out, x, epsilon);
}

} // namespace infinicore::op
